### PR TITLE
Change how api works

### PR DESCRIPTION
**Pull request by mjarkk**
Wow it has been a long time ago i have had so much frustration trying to get something working.  
I've probably re-written all functions arguments, function responses, structs multiple times to get this working due to the borrow checker WHAAAA.  

So i a TL;DR:
- `CompilerProps` now has a `open_file` function requirement
- `Compiler` now has a `open_file` function that we can use cache outputs from `CompilerProps::open_file`
  
Why?  
This makes it supper easy to use the compiler on other places where the file structure might differ from old school IO, like on the web where if we want to use this compiler we have to provide some kind of fake file system or for testing with multiple files we now don't have have the files on the disk.  

A note about why this was so supper frustrating...
I didn't want to clone the file bytes because that would eat memory away supper quickly and at first i thought it was a good idea to maybe use references but after re-writing a lot of things because now everything needs to have a lifetime specifier i came to the conclusion that it's not possible.  
So after a shitload of trail and error i solved it using `Rc<RefCell<dyn CompilerProps>>`.  
The `Rc` is as far as i currently understand it a fancy referent that is more advanced than `&` and doesn't require a lifetime specifier.  
The `RefCell` adds a `.borrow_mut` function that we can use to call functions inside `CompilerProps` that require mutable access.

### CODE REVIEW

**Comment by TheOtterlord** [comment]
It seems that merging in master may have broken something. We have 12 tests failing. 

```
failures:
    compiler::tokenize::tests::enums::test_invalid_inline_enum_in_global
    compiler::tokenize::tests::ifs::test_invalid_if
    compiler::tokenize::tests::ifs::test_invalid_if_else_if
    compiler::tokenize::tests::loops::test_for_incorrect_args_1
    compiler::tokenize::tests::loops::test_for_incorrect_args_2
    compiler::tokenize::tests::loops::test_for_incorrect_args_3
    compiler::tokenize::tests::loops::test_for_no_args
    compiler::tokenize::tests::loops::test_incorrect_while_no_args
    compiler::tokenize::tests::loops::test_loop_incorrect_args
    compiler::tokenize::tests::loops::test_while_incorrect_args
    compiler::tokenize::tests::structs::test_invalid_inline_struct_in_global
    compiler::tokenize::tests::variables::test_variable_global_let_fails
```

If this can be fixed, I think we can merge. Your work looks fantastic and makes the API much better.

File: src/compiler/files.rs [hunk 1]
```diff
@@ -1,30 +1,124 @@
...
+    let mut index = location.index;
+    let mut prev_line: Vec<u8> = vec![];
+    let mut line: Vec<u8> = vec![];
+    let mut next_line: Vec<u8> = vec![];
+    let mut x = 0;
+
+    if index > 0 {
+      while let Some(c) = self.bytes.get(index) {
+        index -= 1;
+        match *c as char {
```
Comment (TheOtterlord) [inline src/compiler/files.rs:73]: This is causing an overflow in some cases

File: src/compiler/files.rs [hunk 2]
```diff
@@ -1,30 +1,124 @@
...
+          '\n' => break,
+          _ => line.insert(0, *c),
+        };
+        x += 1;
+      }
+
+      while let Some(c) = self.bytes.get(index) {
+        index -= 1;
+        match *c as char {
```
Comment (TheOtterlord) [inline src/compiler/files.rs:82]: As is this

**Comment by mjarkk** [comment]
@TheOtterlord Thank you for narrowing down the error I've pushed a fix and that seems to fix the issue.

**Pull request merged**